src/mx_bluesky/hyperion/external_interaction/agamemnon.py

```diff
@@ -0,0 +1,99 @@
+import dataclasses
+import json
+import re
+from typing import TypeVar
+
+import requests
+from dodal.utils import get_beamline_name
+
+from mx_bluesky.common.parameters.components import WithVisit
+from mx_bluesky.common.parameters.constants import GridscanParamConstants
+from mx_bluesky.common.utils.log import LOGGER
+from mx_bluesky.hyperion.parameters.load_centre_collect import LoadCentreCollect
+
+T = TypeVar("T", bound=WithVisit)
+AGAMEMNON_URL = "http://agamemnon.diamond.ac.uk/"
+MULTIPIN_PREFIX = "multipin"
+MULTIPIN_FORMAT_DESC = "Expected multipin format is multipin_{number_of_wells}x{well_size}+{distance_between_tip_and_first_well}"
+MULTIPIN_REGEX = rf"^{MULTIPIN_PREFIX}_(\d+)x(\d+(?:\.\d+)?)\+(\d+(?:\.\d+)?)$"
+
+
+@dataclasses.dataclass
+class PinType:
+    expected_number_of_crystals: int
+    single_well_width_um: float
+    tip_to_first_well_um: float = 0
+
+    @property
+    def full_width(self) -> float:
+        """This is the "width" of the area where there may be samples.
+
+        From a pin perspective this is along the length of the pin but we use width here as
+        we mount the sample at 90 deg to the optical camera.
+
+        We calculate the full width by adding all the gaps between wells then assuming
+        there is a buffer of {tip_to_first_well_um} either side too. In reality the
+        calculation does not need to be very exact as long as we get a width that's good
+        enough to use for optical centring and XRC grid size.
+        """
+        return (self.expected_number_of_crystals - 1) * self.single_well_width_um + (
+            2 * self.tip_to_first_well_um
+        )
+
+
+class SinglePin(PinType):
+    def __init__(self):
+        super().__init__(1, GridscanParamConstants.WIDTH_UM)
+
+    @property
+    def full_width(self) -> float:
+        return self.single_well_width_um
+
+
+def _get_parameters_from_url(url: str) -> dict:
+    response = requests.get(url, headers={"Accept": "application/json"})
+    response.raise_for_status()
+    response_json = json.loads(response.content)
+    try:
+        return response_json["collect"]
+    except KeyError as e:
+        raise Exception(f"Unexpected json from agamemnon: {response_json}") from e
```

Comment:
Nit: We could re-throw the `KeyError` here, after wrapping more context, for better stack trace integrity.
 ```suggestion
         raise KeyError(f"Unexpected json from agamemnon: {response_json}") from e
```
Definitely just a nit, since we propagate up to `update_params_from_agamemnon` which wraps this or the `HTTPError` from `raise_for_status()` (or some other error we don't know about) into an `Exception` anyways, and then swallow the error with a logging statement. So I think we'd lose information about the error type regardless, and would just rely on the f-string reason.